### ISSUE TITLE
corfu-insert and corfu-complete

### ISSUE THREAD

**Issue by xendk** [minad/corfu#93]
This one will need a bit of explanation.

I'm coming from the company-tng frontend, where I've unbound pretty much everything but tab and shift tab, so the popup won't interfere with regular typing (especially the cursor keys and return). Basically taking tab-and-go to the max.

I was considering doing the same for corfu, but then I notice a subtle difference between `corfu-insert` and `corfu-complete`. In css-mode, typing `ba` and `corfu-complete`ing will simply add `ckground` but `corfu-insert` will add `ackground: ;` and place the point just before the `;`.

Now, this has been tripping me up with company, as I would be typing `ba<TAB>:` and end up with `background: :;` because it behaves like `corfu-insert`, and it's not actually obvious which completions will add more than the text displayed in the popup.

Of course, I could just unbind return, and never get tripped up again, but I can see the idea in completing more than just what's in the popup. It's the only way `company-yasnippet` can actually work. Alternatively, I could bind `corfu-insert` to a key more out of the way like `S-<ret>`, but I'd never remember/discover which completions are "magic" in this way and forget to use it.

So, I was wondering if corfu could be more smart about it? I think ("think" because I haven't actually tried it yet) that my dream scenario would be that if the character I type that finishes the completion (the colon in the above example) matches the first character the completion inserts after the preview, that it would not be duplicated.

Oh, and I just remembered. In php-mode with lsp-mode, I can type `$this->get` and get suggestions like `getDelay($subscriberId, $date)` which will preview as `getDelay($subscriberId, $date)`, but if I select it *or* when just typing along, it'll insert `getDelay()` and place the point between the parentheses. Which is kind of the opposite (things get removed instead of added), but works better in practice.

**Comment by minad** [comment]
Did you try the Corfu TNG configuration? See https://github.com/minad/corfu#tab-and-go-completion. From my perspective it already behaves reasonable, but maybe it does not (yet?) have the smartness you are asking for.

> I was considering doing the same for corfu, but then I notice a subtle difference between corfu-insert and corfu-complete. In css-mode, typing ba and corfu-completeing will simply add ckground but corfu-insert will add ackground: ; and place the point just before the ;.

`corfu-complete` performs prefix completion. If the completion is indeed done, it behaves the same as `corfu-insert`. So I don't see a difference there. The difference is when `corfu-complete` only completes the input partially. Then no post completion input is inserted.

> Of course, I could just unbind return, and never get tripped up again, but I can see the idea in completing more than just what's in the popup. It's the only way company-yasnippet can actually work. Alternatively, I could bind corfu-insert to a key more out of the way like S-<ret>, but I'd never remember/discover which completions are "magic" in this way and forget to use it.

Yes, this can be annoying. The magic comes from the backend.

> So, I was wondering if corfu could be more smart about it? I think ("think" because I haven't actually tried it yet) that my dream scenario would be that if the character I type that finishes the completion (the colon in the above example) matches the first character the completion inserts after the preview, that it would not be duplicated.

I don't understand when this would apply. If you select a candidate and type away in TNG style, Corfu already does not add more, in css-mode no `: ;` is added for example. When would the smartness you are describing set in?

> Oh, and I just remembered. In php-mode with lsp-mode, I can type $this->get and get suggestions like getDelay($subscriberId, $date) which will preview as getDelay($subscriberId, $date), but if I select it or when just typing along, it'll insert getDelay() and place the point between the parentheses. Which is kind of the opposite (things get removed instead of added), but works better in practice.

Sounds like a good idea. But these post completion additions are handled by the backend.

Can you be more precise about the exact scenarios and actually try them? I'd appreciate a more concrete proposal, if there is even a need to act. So far I don't see that anything has to be changed, Corfu should already behave in a reasonably smart way I guess.

In any case you can try this alternative implementation of `corfu--done`, which does not add anything after the candidate. You can use the function as an `:override` advice of `corfu--done`. To add smartness you could check for the mode you are using (e.g. css/php) and depending on that call the `:exit-function` which performs post-completion insertion.

~~~ elisp
(defun corfu--done-alternative (str status)
    (undo-amalgamate-change-group corfu--change-group)
    (corfu-quit))
~~~

**Comment by xendk** [comment]
> Did you try the Corfu TNG configuration?

Oh, yes, I started there, should have mentioned that.

> If the completion is indeed done, it behaves the same as corfu-insert.

What do you mean by "done"? I can manage to get tab to insert the same thing as return does. Tab will update the preview, but only when I press return do I get the extra `: ;` part. It makes sense that post-completion input (I'm assuming you mean the `: ;` part) isn't added when completing only part, that would be very unhandy.

I've discovered another oddity. If I type out a completion by hand like `background-clip`, it automatically inserts the `: ;`. If I increase `corfu-auto-delay` it happens after the same delay, unless I type something else first. Seems that it's triggering a completion and the backend adds the extra magic.

> Yes, this can be annoying. The magic comes from the backend.

Well, I'm not against magic, per se, I just want it in a way that doesn't mess with my mojo. :smile: 

> I don't understand when this would apply. If you select a candidate and type away in TNG style, Corfu already does not add more, in css-mode no : ; is added for example. When would the smartness you are describing set in?

In that it *do* add it, but only when it "fits". I know this sounds like border-line mind reading, but... Taking the CSS example, and assuming I've typed `ba` and pressed tab to select `background`: Typing `-` would just insert the `-`, which would trigger another round of completion, typing `%` would just insert the `%` and, as it dosn't match anything, not trigger completion, but typing `:` would let the `: ;` magic happen, but *not* insert another `:`. And the trigger for the magic happening would be that the char typed matched the one inserted directly after `background` by the magic.

But I've digged a bit and found the magic in the CSS completion. It's the `:exit-function`:
```elisp
              :exit-function
              ,(lambda (string status)
                 (and (eq status 'finished)
                      (eolp)
                      prop-table
                      (test-completion string prop-table)
                      (not (and sel-table
                                (test-completion string sel-table)))
                      (progn (insert ": ;")
                             (forward-char -1))))))))))
```

Which is.. Difficult to work with in a generic way. Guess my best option is creating a clone of `css-completion-at-point` that knows when to apply the magic, but how do I determine the char that made corfu decide that the completion was "done"?

But it doesn't explain why `:exit-fuction` is not triggered when I tab to a completion and type on, and why it *is* triggered when I type out a completion in full? The latter seems like a bug to me, but I've only seen it in CSS mode so far.

My current setup, for reference:
```elisp
(use-package corfu
  ;; TAB-and-Go customizations
  :custom
  (corfu-auto t "Enable automatic popup")
  (corfu-auto-prefix 1 "Trigger at one char")
  (corfu-count 30 "Show more candidates")
  (corfu-cycle t "Let suggestions wrap around")
  (corfu-preselect-first nil "Don't select first candidate")

  ;; Use TAB for cycling, default is `corfu-complete'.
  :bind
  (:map corfu-map
        ("TAB" . corfu-next)
        ([tab] . corfu-next)
        ("S-TAB" . corfu-previous)
        ([backtab] . corfu-previous))

  :init
  (corfu-global-mode)
  ;; Clone from GitHup rather than ELPA.
  :straight (:host github :repo "minad/corfu"))
```

**Comment by minad** [comment]
> What do you mean by "done"? I can manage to get tab to insert the same thing as return does. Tab will update the preview, but only when I press return do I get the extra : ; part. It makes sense that post-completion input (I'm assuming you mean the : ; part) isn't added when completing only part, that would be very unhandy.

By "done" I mean that no further completion is possible.

> I've discovered another oddity. If I type out a completion by hand like background-clip, it automatically inserts the : ;. If I increase corfu-auto-delay it happens after the same delay, unless I type something else first. Seems that it's triggering a completion and the backend adds the extra magic.

Good find! This is indeed a bug. Fixed in https://github.com/minad/corfu/commit/5b2612b6d8d4caf7f0377e40eb2668b1cfbcee3d.

> Which is.. Difficult to work with in a generic way. Guess my best option is creating a clone of css-completion-at-point that knows when to apply the magic, but how do I determine the char that made corfu decide that the completion was "done"?

I disagree with your assessment that the current magic of css-mode is unreasonable. But I've only tried it quickly and it seems to work pretty well for me. You may consider writing your own css capf clone or ask upstream to improve the existing exit-function.

> But it doesn't explain why :exit-fuction is not triggered when I tab to a completion and type on, and why it is triggered when I type out a completion in full? The latter seems like a bug to me, but I've only seen it in CSS mode so far.

The behavior is like this now:

1. When you tab to a completion and type on, the exit-function/magic is intentionally disabled, such that what you type is what you get.
2. When you type the candidate in full it adds the magic when the popup is already open.
3. When you type the candidate in full but the popup has not been opened, the magic is not added. 

Before 5b2612b6d8d4caf7f0377e40eb2668b1cfbcee3d the magic was added for both cases 2 and 3 which was indeed a bug.

**Comment by minad** [comment]
So what we could consider to change the magic insertion of each of the following operations - should magic be added or not? I am talking about the default configuration now (corfu-insert, corfu-complete). For the TNG reconfiguration the behavior changes accordingly.

1. When tabbing to a candidate and typing on
2. When typing a candidate exactly and the popup is open
3. When pressing RET (corfu-insert)
4. When pressing TAB (corfu-complete) and completion is done. In this case corfu-complete behaves like corfu-insert.

Would you prefer if the insertion behavior is changed for any of those cases?

**Comment by minad** [comment]
> In that it do add it, but only when it "fits". I know this sounds like border-line mind reading, but... Taking the CSS example, and assuming I've typed ba and pressed tab to select background: Typing - would just insert the -, which would trigger another round of completion, typing % would just insert the % and, as it dosn't match anything, not trigger completion, but typing : would let the : ; magic happen, but not insert another :. And the trigger for the magic happening would be that the char typed matched the one inserted directly after background by the magic.

The suggestion you made here sounds interesting. So if the user has tabbed to a candidate and presses :, the magic should still happen inserting `: ;`  but not the additional colon. It is possible but it goes quite a bit against the design of the `:exit-function` and it would be brittle. The `:exit-function` is executed after completion has finished and can do whatever it wants, Corfu has already left the stage at this point. When a candidate is selected and you type more text, the idea is that the text should be inserted as is. Therefore Corfu exists without executing the `:exit-function`. But I will give it a try to see if it works out.

**Comment by xendk** [comment]
> By "done" I mean that no further completion is possible.

Ah, nevermind what I said about `corfu-complete` vs `corfu-insert`, I seem to have confused myself.

> Fixed in 5b2612b.

But I can't getting it to work, adding the final `p` in `backgroud-cilp` still inserts the `: ;` for some reason.

> I disagree with your assessment that the current magic of css-mode is unreasonable.

Oh, I regret if I've made the impression that I find it unreasonable. I like it. It works excellently with a standard setup of company/corfu. It just ends up clashing a bit with my exotic tastes in completion.

I agree with your assesment that it's pretty much impossible for corfu to do The Right Thing™ as exit-functions might end up doing very different things. I should probably just tweak `css-completion-at-point` to my liking.

As for the four cases... 3 and 4 should do magic, and 2 shouldn't. As for 1... In non-TNG, I think it currently does the right thing. Typing on narrows down the completions or tells you nothing matches. You explicitly trigger the end of completion with tab or ret, and it does it's magic.

But in TNG, tab is simply `corfu-next` which only previews. I'm a bit on the fence and want to eat my cake and have it too. I want it to do magic when I've explicitly tabbed to a candidate and type on. *Unless* the typed character would simply narrow the candidates (tabbing to `background` and entering `-` to get at `background-image` is the obvious example).

Does this make sense?

**Comment by minad** [comment]
Thank you for the feedback.

> But I can't getting it to work, adding the final p in backgroud-cilp still inserts the : ; for some reason.

The issue has been fixed in https://github.com/minad/corfu/commit/5b2612b6d8d4caf7f0377e40eb2668b1cfbcee3d. You probably didn't update properly or there is some other issue with your setup. But to be clear, this only applies if the Corfu popup is not open, since otherwise we would not want to see magic appearing out of the blue! If the popup is open we have case 2 and then magic is added.

> As for the four cases... 3 and 4 should do magic, and 2 shouldn't. As for 1... In non-TNG, I think it currently does the right thing. Typing on narrows down the completions or tells you nothing matches. You explicitly trigger the end of completion with tab or ret, and it does it's magic.

Okay, I agree with you on 3 and 4.  In 1 no magic should happen, and you agree at least in the non-TNG case.

I am not 100% sure about 2 but I like if the magic happens there. I argue that typing the candidate exactly with the popup open is essentially like pressing RET on the candidate. However one could also argue that typing the candidate exactly essentially means that the user ignores the popup and types on, as such no magic should be added.

> But in TNG, tab is simply corfu-next which only previews. I'm a bit on the fence and want to eat my cake and have it too. I want it to do magic when I've explicitly tabbed to a candidate and type on. Unless the typed character would simply narrow the candidates (tabbing to background and entering - to get at background-image is the obvious example).

One could try to distinguish between TNG/non-TNG for case 1. But as I argued it is not possible to do much better. We could add the magic, but then we risk ending up with the double colon. Or we try to fix this afterwards by comparing the inserted text with the last input event, but this is brittle given that the exit-function can do whatever it wants. Corfu/Vertico etc generally avoid adding too much magic on top, so I will not change this.

Anyways I would like to keep the current behavior for now. If I get more reports that the behavior in case 2 is unexpected, I can change this. Regarding 1 I don't see a chance to add magic insertion generally. Of course you can implement this in your config, maybe mode depending.

**Comment by xendk** [comment]
> But to be clear, this only applies if the Corfu popup is not open, since then we don't want to see magic appearing out of the blue! If the popup is open we have case 2 and then magic is added.

Aaah, that's why, I'm running with a extremely short delay. 

>  In 1 no magic should happen, and you agree with me about that at least in the non-TNG case.

Well, the magic does happen, but only when explicitly selecting a completion.

> I am not 100% sure about 2 but I like if the magic happens there. I argue that typing the candidate exactly with the popup open is essentially like pressing RET on the candidate. However one could also argue that typing the candidate exactly essentially means that the user ignores the popup and types on, as such no magic should be added.

Ah, but I'll argue that it's exactly here you have the sharp edges between TNG and non-TNG. In TNG you don't have direct control over when the popup is shown, so a moments' hesitation while typing can make it pop up on something you don't want magic happening on. My point of view is that if you (in non-TNG) explicitly have asked for completions, and then proceed to type it out by hand, there's probably something about the suggestions you don't like, and adding magic is unlikely to make you happier.

> but then we risk ending up with the double colon

Well, yeah, I'm tempering my hopes for mind-reading capabilities and thinking I might be able to adjust to it. Think part of the problem was that I didn't quite understand it, but now I've looked through the css-mode code, and it stands much clearer.

So I think it makes sense to add magic when typing away, if a candidate has been selected and completion is done. Else I figure stuff like cape-abbrev wouldn't work in TNG mode?

**Comment by minad** [comment]
> My point of view is that if you (in non-TNG) explicitly have asked for completions, and then proceed to type it out by hand, there's probably something about the suggestions you don't like, and adding magic is unlikely to make you happier.

Note that you can always press C-g to quit completion. Your argument is not convincing. If you type the suggestion fully out by hand, it obviously is what you want. 

But really the question of adding magic is debatable, I am not yet decided and I have to think a bit more about it. An argument to not add the magic is to make the behavior more symmetric with the no-match-behavior with `corfu-quit-no-match=t`.

> So I think it makes sense to add magic when typing away, if a candidate has been selected and completion is done. Else I figure stuff like cape-abbrev wouldn't work in TNG mode?

Yes, tabbing and typing away does not work for `cape-abbrev`. You have to confirm explicitly with RET. It may not feel comfortable like this, but it also shouldn't be too hard to adapt your muscle memory. To repeat myself *generally adding magic when typing away is not possible* (because of the double colon problem etc). Of course you can still hack Corfu in your configuration and add mode-specific tweaks.

**Comment by xendk** [comment]
> Note that you can always press C-g to quit completion.

Yes, but that argument only holds for the non-TNG case. I don't know if you prefer to keep behaviour consistent between the two modes, and consequently pick a solution that works for both, or make them diverge, but I'll reiterate my point in more detail, as I think it's rather important: In auto mode you don't control when the popup is shown, so having different behaviours depending on its visibility is bad UX. You're putting a cognitive load on the user, who has to keep track of whether the popup is shown. Being able to back out if it doesn't help as there's always the chance of it showing up in the time it takes for the user to notice its absence to them hitting the key.

> (EDIT: Since you mentioned cape-dabbrev, this is a strong point to keep the current behavior. When typing the abbrev exactly you want it to expand!)

Not in TNG I wont! I love dabbrev, I use it with company-tng in text buffers and code comment, being able to type the first few characters of a long word, tab to completion and typing on feels like having a smartphone keyboard dictionary built into emacs. It's "da bomb". But I *don't* want it to fix `some` to `somewhere` *unless* I've tabbed to it.

> Yes, tabbing and typing away does not work for cape-abbrev.

Why not? It works for `company-dabbrev` with `company-tng` just fine, in my opinion. Yes, there's the double colon issue, but I think the ergonomics of tap'n'typing with dabbrev *far* outweigh the (slight) problem with one mode (css-mode is the only one that I've encounted).

The issue with having to use enter (apart from turning an, often, one key operation into a two key one) is that it splits completion in, not two, but three different groups I have to adjust my muscle memory for: Those which are tab-and-type (elisp and lsp (which *does* have magic, must not be using `:exit-function` then)), those with tab-and-enter-and-type (dabbrev, yasnippet if someone implements it) and tab-and-type-but-with-magic-with-enter (css-mode).

I'd rather prefer them all to be tab-and-type-with-magic and deal with the odd case by working on my muscle memory for that language, or creative hacking.
